src/file_storage/policy.c

```diff
@@ -0,0 +1,229 @@
+    uint64_t lltmp;
+
+    meminfo_t meminfo;
+
+    pch = strtok(str, "%");
+    if(pch !=NULL) {
+        /* Check if the input string is a percentage number */
+        swap->mem_quota_type = 1;
+        swap->mem_quota.percent = strtof(pch, NULL) / 100.0;
+    } else if (sscanf(str, "%f %n", &ftmp, &len) && !str[len]) {
+        /* Check if the input string is a float number */
+        swap->mem_quota_type = 1;
+        swap->mem_quota.percent = ftmp;
+    } else if((pch=strtok(str, "kKmMgGtT")) != NULL) {
+        /* Check if the input string is a real value with quantity */
+        swap->mem_quota_type = 0;
+        ftmp = strtof(pch, &pch);
+        // Check if the last char is 'B'(Byte) or 'b'(bit)
+        if(*(pch+1) == 'b')
+            ftmp = ftmp / 8;
+        // Check the quantity.
+        if(*pch == 'k' || *pch == 'K') {
+            ftmp = ftmp / 2e10;
+        } else if (*pch == 'm' || *pch == 'M') {
+            ;
+        } else if (*pch == 'g' || *pch == 'G') {
+            ftmp = ftmp * 2e10;
+        } else if (*pch == 't' || *pch == 'T') {
+            ftmp = ftmp * 2e20;
+        }
+        // If the quota is higher than the full memory capacity,
+        // Just set the quota to 100%
+        meminfo = parse_meminfo();
+        if(ftmp > ((float) meminfo.MemTotalMiB)) {
+            swap->mem_quota_type = 1;
+            swap->mem_quota.percent = 1.0;
+        }
+        swap->mem_quota.MB = ftmp;
+    }
+}
+
+void disk_quota_parser(char* str, struct swap_config* swap)
+{
+    char *pch;
+    float ftmp;
+
+    pch = strtok(str, "kKmMgGtTpP");
+    if(pch == NULL) {
+        swap->disk_quota_MB = -1.0;
+    } else {
+        // Round to an approx int
+        ftmp = strtof(pch, &pch);
+        // Check if the last char is 'B'(Byte) or 'b'(bit)
+        if(*(pch+1) == 'b')
+            ftmp = ftmp / 8;
+        // Check the quantity.
+        if(*pch == 'k' || *pch == 'K') {
+            ftmp = ftmp / 2e10;
+        } else if (*pch == 'm' || *pch == 'M') {
+            ;
+        } else if (*pch == 'g' || *pch == 'G') {
+            ftmp = ftmp * 2e10;
+        } else if (*pch == 't' || *pch == 'T') {
+            ftmp = ftmp * 2e20;
+        } else if (*pch == 'p' || *pch == 'P') {
+            ftmp = ftmp * 2e30;
+        }
+        swap->disk_quota_MB = ftmp;
+    }
+}
+
+static int default_when_policy(uint64_t size_MB)
+{
+    meminfo_t meminfo = parse_meminfo();
+
+    if(meminfo.MemAvailableMiB < size_MB)
+        return 1;
+    else
+        return 0;
+}
+
+static int percent_when_policy(float threshold, uint64_t size_MB)
+{
+    meminfo_t meminfo = parse_meminfo();
+
+    if((meminfo.MemAvailablePercent - (size_MB / meminfo.MemTotalMiB)) < (1 - threshold))
+        return 1;
+    else
+        return 0;
+
+}
+
+static int value_when_policy(float threshold, uint64_t size_MB)
+{
+    meminfo_t meminfo = parse_meminfo();
+
+    if(meminfo.MemAvailableMiB  < (size_MB + meminfo.MemTotalMiB - threshold))
```

Comment:
Yes, we can definitely add a memory usage keeper for `ls_add()` and `ls_remove()`. But the problem is, and also the reason why I didn't implement this way is: we are not actually allocating a new memory region when we add `od` to `ls` (`ls_remove()` really releases the memory), but we allocate the memory for `od` before the bulk transfer, and assign the `od->data` as the bulk transfer destination, then we attach this `od`'s ptr to the `ls`. So if we really want to keep track of the memory usage to prevent the overflow, we might already got segfault before the checking in `ls_add()`

Reply:
We can address the order of operations by making a memory tally for local storage and doing setters and getters for it. In reality, running out of physical memory won't segfault, it will start swapping into virtual memory.